ledger/metrics/src/main/scala/com/daml/telemetry/TelemetryContext.scala

```diff
@@ -0,0 +1,235 @@
+import io.opentelemetry.api.trace.Span
+import io.opentelemetry.context.Context
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes
+
+import scala.concurrent.Future
+import scala.util.{Failure, Success}
+
+trait TelemetryContext {
+
+  /** Sets or replaces the value of attribute `key` to `value`.
+    */
+  def setAttribute[T](key: AttributeKey[T], value: T): TelemetryContext
+
+  /** Creates a new span and runs the computation inside it.
+    * The new span has its parent set as the span associated with the current context.
+    * A new context containing the new span is passed as parameter to the computation.
+    *
+    * @param spanName   the name of the new span
+    * @param kind       the kind of the new span
+    * @param attributes the key-value pairs to be set as attributes to the new span
+    * @param body       the computation to be run in the new span
+    * @return the result of the computation
+    */
+  def runFutureInNewSpan[T](
+      spanName: String,
+      kind: SpanKind,
+      attributes: (AttributeKey[String], String)*
+  )(
+      body: TelemetryContext => Future[T]
+  ): Future[T]
+
+  /** Creates a new span and runs the computation inside it.
+    * The new span has its parent set as the span associated with the current context.
+    * A new context containing the new span is passed as parameter to the computation.
+    *
+    * @param spanName the name of the new span
+    * @param kind the kind of the new span
+    * @param attributes the key-value pairs to be set as attributes to the new span
+    * @param body the computation to be run in the new span
+    * @return the result of the computation
+    */
+  def runInNewSpan[T](
+      spanName: String,
+      kind: SpanKind,
+      attributes: (AttributeKey[String], String)*
+  )(
+      body: TelemetryContext => T
+  ): T
+
+  /** Runs the computation inside an OpenTelemetry scope.
+    *
+    * This is used to set the tracing metadata in the gRPC local thread context, so it
+    * can be accessed and used by the OpenTelemetry instrumentation to create complete traces.
+    *
+    * It should be used around gRPC calls, to ensure that the tracing metadata is correctly used
+    * and transferred.
+    *
+    * @param body the computation to be run in the Telemetry scope
+    * @return the result of the computation
+    */
+  def runInOpenTelemetryScope[T](body: => T): T
+
+  /** Encode the metadata of the context in a key-value map.
+    *
+    * Typically, metadata is encoded in a map to allow transporting it across process boundaries.
+    * Originally, it has been created to carry tracing metadata across boundaries, and
+    * to create complete traces.
+    *
+    * @see [[com.daml.telemetry.Telemetry.contextFromMetadata(java.util.Map)]]
+    */
+  def encodeMetadata(): jMap[String, String]
+
+}
+
+object TelemetryContext {
+
+  /** Sets the attributes extracted from the given gRPC `MethodDescriptor`.
+    */
+  def setGrpcAttributes(
```

Comment:
There are a few methods like this that are not being used in this repo. I'm not sure if I have to address this. Maybe in the next PR.

Reply:
Yup, `TelemetryContext.runInOpenTelemetryScope`, `TelemetryContext.encodeMetadata`, `TelemetryContext.setGrpcAttributes` (this one is already removed in the next PR). As discussed offline, we're going to keep them most likely for a short period of time.

Reply:
Updated the above comment with the missing info.